mlir/lib/Catalyst/Transforms/InlineNestedModules.cpp

```diff
@@ -0,0 +1,493 @@
+ *       func.call @baz()
+ *     }
+ *     ```
+ * 5. Cleanup: remove the catalyst.fully_qualified_name attribute
+ *
+ */
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Catalyst/Transforms/Passes.h"
+#include "Catalyst/Transforms/Patterns.h"
+#include "Gradient/IR/GradientInterfaces.h"
+#include "Mitigation/IR/MitigationOps.h"
+
+#include <deque>
+
+using namespace mlir;
+
+namespace {
+
+void getFullyQualifiedName(SymbolOpInterface symbol, const Operation *op,
+                           std::deque<FlatSymbolRefAttr> &hierarchy)
+{
+    auto stringRef = symbol.getNameAttr();
+    auto flatSymbolRef = SymbolRefAttr::get(stringRef);
+    hierarchy.push_front(flatSymbolRef);
+
+    auto parent = symbol->getParentOp();
+    auto parentSymbol = parent ? dyn_cast<SymbolOpInterface>(parent) : nullptr;
+    auto parentIsLimit = parent == op;
+    auto isValidParent = parent && parentSymbol && !parentIsLimit;
+    if (!isValidParent)
+        return;
+
+    getFullyQualifiedName(parentSymbol, op, hierarchy);
+}
+
+SymbolRefAttr getFullyQualifiedNameUntil(SymbolOpInterface symbol, const Operation *op)
+{
+    auto symbolTable = symbol->getParentOp();
+    assert(symbolTable->hasTrait<OpTrait::SymbolTable>() &&
+           "symbolTable must have OpTrait::SymbolTable");
+
+    std::deque<FlatSymbolRefAttr> hierarchy;
+    getFullyQualifiedName(symbol, op, hierarchy);
+    assert(hierarchy.size() > 0 && "At least one symbol is expected here.");
+    bool inNamelessModule = hierarchy.size() == 1;
+    if (inNamelessModule) {
+        return SymbolRefAttr::get(symbol);
+    }
+
+    FlatSymbolRefAttr root = hierarchy.front();
+    hierarchy.pop_front();
+    SmallVector<FlatSymbolRefAttr> hierarchy_vector(hierarchy.begin(), hierarchy.end());
+    auto ctx = symbol.getContext();
+    return SymbolRefAttr::get(ctx, root.getValue(), hierarchy_vector);
+}
+
+struct AnnotateWithFullyQualifiedName : public OpInterfaceRewritePattern<SymbolOpInterface> {
+    using OpInterfaceRewritePattern<SymbolOpInterface>::OpInterfaceRewritePattern;
+
+    /// This overload constructs a pattern that matches any operation type.
+    AnnotateWithFullyQualifiedName(MLIRContext *context, Operation *root)
+        : OpInterfaceRewritePattern<SymbolOpInterface>::OpInterfaceRewritePattern(context),
+          _root(root)
+    {
+    }
+
+    LogicalResult match(SymbolOpInterface symbol) const override;
+    void rewrite(SymbolOpInterface symbol, PatternRewriter &rewriter) const override;
+
+    const Operation *_root;
+};
+
+static constexpr llvm::StringRef fullyQualifiedNameAttr = "catalyst.fully_qualified_name";
+
+LogicalResult AnnotateWithFullyQualifiedName::match(SymbolOpInterface symbol) const
+{
+    auto hasQualifiedName = symbol->hasAttr(fullyQualifiedNameAttr);
+    return hasQualifiedName ? failure() : success();
+}
```

Comment:
Just a suggestion since this pertains to style, but I think these would be more readable inlined into the class when they are reasonably short, thereby reducing the number of top-level constructs in this file (1 instead of 3 per pattern, especially given that there is no header-source separation for these classes).